### PR TITLE
Initial implementation (& discussion)

### LINKED ISSUE #1: API discussion
A ticket to discuss the API.

The other day I was thinking something like this could work:

```purescript
ajax :: forall e. AjaxRequest -> Aff (ajax :: AJAX | e) AjaxResponse

type AjaxRequest = {
  url :: String,
  method :: Method,
  headers :: StrMap String,
  content :: Content,
  ...
}

type AjaxResponse = {
  headers :: StrMap String
  ...
}

ajaxRequestDefaults :: String -> AjaxRequest
```

Then a separate module could define combinators, such as:

```purescript
GET :: AjaxRequest -> AjaxRequest

contentType :: MimeType -> AjaxRequest -> AjaxRequest
...
runAjax :: String -> (AjaxRequest -> AjaxRequest) -> Aff (ajax :: AJAX | e) AjaxResponse
runAjax url f = ajax (f (ajaxRequestDefaults url))
```

So you'd write code like:

```purescript
runAjax "http://foo.com" $ GET >>> contentType applicationJson
```

But I'm sure there are better APIs. Let's discuss here if you have any ideas.

### PR DESCRIPTION
**Pull request by garyb**
I haven't tested this at all yet, so it's definitely not ready, but there are a few things I thought it'd be worth talking about at this point.

1. How do we want to deal with the [various content types](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#send%28%29) that can be sent?
2. What do we want for the response? Is plain text enough? What about reading response headers?
3. Is the free monad overkill?

I spent a while trying things out for point 3, including a couple of ways of implementing it without using `purescript-free`, but it did seem silly not to take advantage of what we already have there. However, I'm acutely aware of the amount of machinery being used to enact a pretty trivial task. But then I guess it doesn't have to be super fast, as it's unlikely that people will need to make a lot of requests together, and even if they do the `AjaxRequest` type is still available to be used directly if so desired.

Any feedback is welcome, but I am intending to check things over and actually try running the code before calling it done! So really it's more the questions about the implementation that I'm looking for input on right now.

Resolves #1

### CODE REVIEW

**Comment by paf31** [comment]
Looks nice to me. It's nice to see `Free` being used some more.

Thinking out loud, maybe something like `RebindableSyntax` for `>>=` would let you constrain things like limiting to one method, or one URL.

Have you considered using something like `purescript-options` for the AJAX request object?

Maybe a bunch of newtypes and a type class would be a nice way to handle the various content types? `newtype PlainText`, `newtype JSON`, etc.

For the response, something like `Foreign` seems like a decent default for the low-level API, which you could then layer abstractions on top of.

**Comment by garyb** [comment]
Even though I raised the point originally in the API discussion I became less concerned about setting the method or URL more than once - I figured the compositional advantage of not restricting that might be worth it. For example, you might want to make most of a request in one place, and then override the method or endpoint later on before running it, while keeping headers or content or something.

I did end up putting something in for the content types, but it depends on [these additions to `purescript-dom`](https://github.com/purescript-contrib/purescript-dom/pull/6) - which I don't know if we want to add necessarily, but it seemed sensible to have canonical foreign types for those things.

JSON isn't a "native" type for XHR, but it probably would be nice to have a combinator that handles that, but then I don't know what type to use for that either! Could be `forall a. { | a}`, could be `Foreign`, could be `Json` from `purescript-argonaut` etc. so again I left that out for now. `Json` would be the "best" solution, but pulling argonaut in to this is probably overkill.

I hadn't even thought of `purescript-options` actually, but yeah, that would probably work.

`Foreign` probably is the way to go for the response too, given it can be one of a number of types. Perhaps there could be two ways of running a request, one returning `Foreign` and the other `(Foreign, [Header])` or something like that.

File: src/Network/Affjax/Request.purs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+-- | The types of data that can be set in an AJAX request.
+data Content a
+  = ArrayViewContent (ArrayView a)
+  | BlobContent Blob
+  | DocumentContent Document
+  | TextContent String
+  | FormDataContent FormData
+
+-- TODO: probably not this? Do we want to deal with other responses, include headers, etc?
+newtype AjaxResponse = AjaxResponse String
```
Comment (jdegoes) [inline src/Network/Affjax/Request.purs]: Yeah, we'll somehow want to support `responseBody`, `responseText`, and `responseXML`, since I think those are the 3 baked into `XMLHttpRequestObject`, together with headers and status code. If this is just a record, it's easy enough for users to extract out the appropriate field of interest, so including "too much information" (e.g. headers) is probably OK.

File: src/Network/Affjax/Request.purs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  , content :: Maybe (Content a)
+  , username :: Maybe String
+  , password :: Maybe String
+  }
+
+-- | A HTTP method name: `GET`, `POST`, etc.
+newtype MethodName = MethodName String
+
+-- | The types of data that can be set in an AJAX request.
+data Content a
```
Comment (jdegoes) [inline src/Network/Affjax/Request.purs]: Are these all (& only) the content types supported by AJAX?

File: src/Network/Affjax/Request.purs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    var xhr = new XMLHttpRequest();
+    xhr.open(method, url, true, username, password);
+    for (var i = 0, header; header = headers[i]; i++) {
+      xhr.setRequestHeader(header.header, header.value);
+    }
+    xhr.onerror = function (err) {
+      errback(err)();
+    };
+    xhr.onload = function () {
+      if (xhr.status === 200) callback(xhr.response)();
```
Comment (jdegoes) [inline src/Network/Affjax/Request.purs]: The status codes can sometimes have semantic (e.g. Pending, redirects, etc.), so we'll want to be sure to bubble them up.

File: src/Network/Affjax/DSL.purs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+affjaxN :: forall c. Natural (AffjaxRequestF c) (State (AjaxRequest c))
+affjaxN (SetURL url a) = const a <$> modify (_ { url = url })
+affjaxN (SetMethod method a) = const a <$> modify (_ { method = method })
+affjaxN (AddHeader header a) = const a <$> modify (\req -> req { headers = header : req.headers })
+affjaxN (SetContent content a) = const a <$> modify (_ { content = content })
+affjaxN (SetUsername username a) = const a <$> modify (_ { username = username })
+affjaxN (SetPassword password a) = const a <$> modify (_ { password = password })
+
+-- | Runs the DSL, producing an `AjaxRequest` object.
+affjaxRequest :: forall c a. AffjaxRequest c a -> AjaxRequest c
```
Comment (jdegoes) [inline src/Network/Affjax/DSL.purs]: I suggest we expose a `runAffjax` which not only produces the request but also runs it using the `ajax` method, because it will be very common to one to do both (rather than produce a request data structure).

File: src/Network/Affjax/Request.purs (hunk 3 above)
Comment (cryogenian) [inline src/Network/Affjax/Request.purs]: :+1:

File: src/Network/Affjax/Request.purs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  { url :: String
+  , method :: MethodName
+  , headers :: [Header]
+  , content :: Maybe (Content a)
+  , username :: Maybe String
+  , password :: Maybe String
+  }
+
+-- | A HTTP method name: `GET`, `POST`, etc.
+newtype MethodName = MethodName String
```
Comment (cryogenian) [inline src/Network/Affjax/Request.purs]: why not? 
```purescript
data MethodName = GET | POST | PUT | UPDATE | DELETE | MethodName String
```

File: src/Network/Affjax/DSL.purs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+affjaxRequest = (`execState` defaultRequest) <<< runFreeCM affjaxN
+
+-- | Take a standard HTTP verb and allow it to be used as an AJAX request
+-- | method.
+methodNameFromVerb :: Verb -> MethodName
+methodNameFromVerb = MethodName <<< show
+
+-- | Sets the URL for a request.
+url :: forall c. String -> AffjaxRequest c Unit
+url url = liftFC (SetURL url unit)
```
Comment (jdegoes) [inline src/Network/Affjax/DSL.purs]: I'd also like to see combinators like `get "http://foo.com"` which set the method and the URL at the same time, so you could write code like:

```purescript
do get "http://foo.com"
   ...
```

**Comment by cryogenian** [comment]
There is an option to set expected response type i.e. 
```purescript
class Responsable a where 
  toStr :: a -> String 

request = do
  ...
  expected Json
```

**Comment by cryogenian** [comment]
there is an option to set expected type 
```purescript
foreign import data RawResponse :: * 
 
class Response a where 
  to :: RawResponse -> a

instance rJson :: Response Json where 
  ...

-- then add response type to AffjaxRequest
```

Or we can use 
```purescript
data Response a = ApplicationJson String | XML String | RawText String
```
and check response type in ffi function

File: src/Network/Affjax/Request.purs (hunk 2 above)
Comment (garyb) [inline src/Network/Affjax/Request.purs]: Yeah, that covers all the natively acceptable types now. It didn't in my original commit.

File: src/Network/Affjax/Request.purs (hunk 3 above)
Comment (garyb) [inline src/Network/Affjax/Request.purs]: Ah yep, good point. I can include that in the response record mentioned elsewhere in this thread.

File: src/Network/Affjax/DSL.purs (hunk 4 above)
Comment (garyb) [inline src/Network/Affjax/DSL.purs]: [Like this](https://github.com/slamdata/purescript-affjax/blob/684226309f80676920312295213c33ded6624fdb/src/Network/Affjax.purs#L8-L9)? :)

**Comment by jdegoes** [comment]
@cryogenian The 2nd approach won't work, because the client has to "request" the type of response -- the browser will lazily compute it if possible (for example, if you request XML, it will be parsed as XML, but not if you request text).

The 1st approach might work for this. Then users call the `to` function (hopefully by some better name :smile: ), to convert from the response to something that is provably obtainable from the XMLHttpRequest (i.e. responeBody, responseXML, responseTxt). Argonaut's JSON type could implement such a thing.

File: src/Network/Affjax/Request.purs (hunk 5 above)
Comment (garyb) [inline src/Network/Affjax/Request.purs]: I wanted to avoid defining these at all originally, and just use the verbs from `purescript-http` until I saw the requirement for custom verbs. The `method` combinator does accept the `purescript-http` verbs at the moment at least.

Perhaps I should make a PR to allow custom verbs there instead.

File: src/Network/Affjax/DSL.purs (hunk 4 above)
Comment (jdegoes) [inline src/Network/Affjax/DSL.purs]: Yes, precisely! :smile:

File: src/Network/Affjax/DSL.purs (hunk 6 above)
Comment (garyb) [inline src/Network/Affjax/DSL.purs]: :+1:

File: src/Network/Affjax/DSL.purs (hunk 4 above)
Comment (garyb) [inline src/Network/Affjax/DSL.purs]: Do you think it belongs in `Affjax.DSL` instead then, or is it fine where it is?

File: src/Network/Affjax/Request.purs (hunk 5 above)
Comment (jdegoes) [inline src/Network/Affjax/Request.purs]: The design of http is not very good and it doesn't look maintained, so I'd suggest keeping this self-contained and doing as Maxim suggested.

**Comment by paf31** [comment]
Would it be possible to just tag the request and response types with the type of the response data to ensure that they're compatible?

So a `Request Blob` would result in a `Response Blob` or whatever, and then you provide smart constructors to create requests with a finite set of acceptable content types.

File: src/Network/Affjax/DSL.purs (hunk 4 above)
Comment (jdegoes) [inline src/Network/Affjax/DSL.purs]: I would expect it to be in DSL, but am curious why you put it in its own module?

**Comment by garyb** [comment]
The request content type has no connection to the response unfortunately as the request content-type header does not determine the response. They don't even have the same set of potential types (supposedly there is an XHR JSON response, although I don't know how you're supposed to get one as `application/json` content-type response headers don't trigger it, along with some moz-specific stuff).

**Comment by paf31** [comment]
Sorry, I was thinking of the Accept header.

File: src/Network/Affjax/DSL.purs (hunk 4 above)
Comment (garyb) [inline src/Network/Affjax/DSL.purs]: I figured it was like the "main" affjax method as it ties DSL and requests together - the DSL module purely builds the requests, and the Request module handles the execution of them.

If we could do it I would re-export the contents of both `DSL` and `Request` submodules from `Network.Affjax` actually, but I've not gotten around to adding that to the compiler yet.

**Comment by garyb** [comment]
Oh, maybe I misunderstood too, `AjaxRequest` is already parameterised with the send-content-type, did you mean just adding another parameter to cover the responses also?

**Comment by paf31** [comment]
Yes - maybe it doesn't make sense, I've been kind of dipping in and out of the conversation. But it seems to make sense to be able to say "this is what I expect in the response, fail if it's anything else". You can handle it when you get the response using `foreign`, `argonaut`, or whatever, but this seems to assert a little extra type safety between request and response. It changes the meaning of the errors to "the server is probably doing something wrong".

**Comment by jdegoes** [comment]
Yeah, request and response types can't be tied together, but it would be nice to tie request / response encoders & decoders to mime types, so you can get those errors like @paf31 mentioned.

**Comment by jdegoes** [comment]
```
class AjaxRequestable a where
  mimeType :: MimeType
  to :: a -> Foreign

class AjaxResponsable a where
  mimeType :: MimeType
  from :: Foreign -> Either Error a
```

**Comment by garyb** [comment]
:+1:

Could make it `forall req res. (IsForeign res) => AjaxRequest req res` perhaps then?

**Comment by garyb** [comment]
Oh, or that :)

File: src/Network/Affjax/Request.purs (hunk 5 above)
Comment (garyb) [inline src/Network/Affjax/Request.purs]: What's the problem with it? Seems the readme is a little out of date, but it's just a collection of types for the most part: https://github.com/joneshf/purescript-http/blob/master/src/Network/HTTP.purs

I only ask as I'm using the `Header` types from there as well at the moment. I don't mind re-implementing both if you'd rather I did it that way though.

Comment (jdegoes) [inline src/Network/Affjax/Request.purs]: 1. `StatusCode` does not support custom.
2. `Verb` does not support custom.
3. `Header` is not well-typed; the value is a `String` but much more precise types are possible (values have a lot of structure, they are almost never free-form text!).
4. `Show` is abused; `Show` instances should render as REPL-friendly copy/paste code. Instead, functions should be used for rendering these things.
5. Not all known verbs, codes, header heads are implemented.

I do think we need a good http-like library, one which is type safe and maintained. That may be too big a project for now (but open to whatever your thoughts are on the issue).

**Comment by garyb** [comment]
I'll try and address the remaining response-related things tomorrow, I've made the `Header` and `Verb` stuff local to affjax now.

`Header` is just as bad as it was before, if not worse considering I've not provided a set of default `HeaderHead`s, but I figured for now I'd just go with something simple that will work until we have a proper HTTP library. Maybe this is still wrong though, as I guess to get correctly typed `Header` values we can't separate the value from the `HeaderHead`.

**Comment by jdegoes** [comment]
Yes, the `Header` must be a sum type, where each term encodes the type-safe values possible by its constructor. So maybe the right thing to do in the short term is something quick & dirty, e.g.:

```purescript
data Header = Unknown String String -- or data Header = Header String String
```

A well-done, type-safe HTTP library would be useful here as well as for any node.js HTTP libraries.

Oh, forgot to mention another problem with `http`, that is, it doesn't separate `RequestHeader` from `ResponseHeader`, and the two are actually quite different; they share a few terms but are conceptually and semantically distinct. Most types of request headers don't make sense in responses, and visa versa.

**Comment by garyb** [comment]
I'm having some difficulty with switching over to using `AjaxRequestable`.

1. If I change the `content` property to `a`
2. create a `NoContent` value with `AjaxRequestable` instance
3. update `defaultRequest` to use that and be typed as `AjaxRequest NoContent`

`affjaxRequest` doesn't work out anymore, as the type in `State` is expected be `AjaxRequest NoContent`.

However, if I keep `content` as `Maybe a` it becomes annoying to use, as making a simple call without setting some content results in errors like `No instance found for Network.Affjax.Request.AjaxRequestable u15626`.

**Comment by garyb** [comment]
One option would be to make `content`'s type `Foreign`, but that doesn't seem like a very good idea.

At first I wanted to implement `AjaxRequestable`'s `to` as being `a -> Content`, which would also solve this I think, but I ran in to a difficulty there as well.

Currently `Content`'s `a` parameter is for the `ArrayViewContent` constructor, I was going to make use `Exists ArrayView` so we could drop the parameter from `Content`, but @paf31 pointed out on IRC that allows any kind of nonsense - which is also true with `Content a` at the moment. Perhaps having `ArrayViewContent` use a sum type as its argument that restricts the possible set to sensible types is the way to go, although not entirely convenient.

**Comment by garyb** [comment]
Ok, found a middle ground. I introduced a foreign `AjaxContent` type, which is basically just `Foreign` but with more meaning at least.

The `AjaxRequestable` class has a bunch of default instances for the "low level" content types to allow those to be posted, I'm not 100% sure on the mime types provided for those, but the idea is they will make it easier to produce `AjaxRequestable` instances for higher level types rather than necessarily be relied on directly.

**Comment by garyb** [comment]
I'm starting a new branch, I think using `purescript-options` might be a better approach.
